### ISSUE TITLE
[Memory Engine] Write Pipeline: internal data processing and storage

### ISSUE THREAD

**Issue by decster** [apache/incubator-doris#3469]
This includes how the memory engine handles data write/load quests, write tx WAL, apply write ops etc.

**Comment by vagetablechicken** [comment]
### Data Flow
1.  User data(json format) -> PartialRow
    1. PartialRow(Type,...), Type: INSERT/UPSERT/DELETE/UPDATE
    1. In POC: PartialRow(cells[0],...), cells[0].isset == 1 means DELETE. Don't distinguish between INSERT and UPDATE.
1. PartialRowBatch ---transport---> ... --> DeltaWriter::write(PartialRow row)

We focus on DeltaWriter.

#### DeltaWriter
About DeltaWriter, needs a new class, effects similar to MemTable, temporarily named WALMemTable. Flush WAL when DeltaWriter closed.
1. DeltaWriter prepare works
1. DeltaWriter::write(PartialRow row)
1. DeltaWriter::close()
    1. WALMemTable::flush()
    1. WAL file format(the simplest): one line is a partial row, saved in {data_dir}/wal/{shard_id}/{tablet_id}/{schema_hash}/, filename: {txn_id}.wal
1. DeltaWriter::close_wait()
    1. Apply rows to SubTablet (If meet errors, e.g. can't find key for DELETE/UPDATE, return error)
    1. commit_txn()

**Comment by decster** [comment]
I think we can attach a set of default values to each writetx, so if update became insert, we can fillin default value automatically.

The default value can come from create table statement(each column can have a default value), and also come from merge into statement, e.g.

```
MERGE sales.category t 
    USING sales.category_staging s
ON (s.category_id = t.category_id)
WHEN MATCHED
    THEN UPDATE SET 
        t.amount = s.amount
WHEN NOT MATCHED BY TARGET 
    THEN INSERT (category_id, category_name, amount)
         VALUES (s.category_id, 'default_name', s.amount)
WHEN NOT MATCHED BY SOURCE 
    THEN DELETE;
```

**Comment by vagetablechicken** [comment]
### WAL file format
WAL file will save multiple partial rows. One write TX's WAL may be split by size => generates multiple {txn_id}_{segment_id}.wal

For the related data structs, such as ParitalRowBatch, refer to https://github.com/decster/choco/blob/b644430e4e540da8ce0a6d13f07223c818706425/src/choco/partial_row_batch.h#L11

A WAL file has the same struct as PartialRowBatch, shown below. WAL file size could be determined by config::write_buffer_size.

![image](https://user-images.githubusercontent.com/24697960/82416937-715a8980-9aad-11ea-8fb4-90a3fabb2176.png)

The schema part, we could reuse the definition in segment proto.
https://github.com/apache/incubator-doris/blob/d2d95bfa841245d734bcd416b4c429ebe98b8321/gensrc/proto/segment_v2.proto#L158

**Comment by vagetablechicken** [comment]
### Prepare Phase Detail Design
#### Sender BE:
```
Get batch from json_scanner
For partialrow in batch:
    // may missing key cols(partition cols or distributed cols), should fillin default value.
    // If we want to support default values for each write txn, those default values need to 
    // be pass to receiver BE(e.g. PTabletWriterOpenRequest adds a field). And no need to 
    // modify rows here, coz we only need key cols here. Filling all key & value cols 
    // could exec in commit phase. 
    calc_partition_and_tablet(partialrow, &channel, &tablet_id)
    channel.add_row(partialrow, tablet_id)
```

#### Receiver BE:
```
Get batch from sender BE
For partialrow, tablet_id of batch:
    get delta_writer by tablet_id
    delta_writer.write(partialrow)
```

 ##### DeltaWriter(DeltaWriter has two mode, one for disk engine, one for memory engine. Sets up the mode upon initialization)
```
// change arg Tuple* to void*
// disk engine MemTable will convert it to Tuple
// memory engine MemTable will convert it to PartialRow
DeltaWriter::insert(void* row):
    mem_table.insert(row)
    if memory_usage > config:
        _flush_memtable_async()
        _reset_mem_table()
```
##### WTDMemTable(WriteTxnData MemTable, flush WriteTxnData files)
Member var: _buffer, ...
```
WTDMemTable::insert(void* row):
    append(_buffer, row, sizeof(uint64_t)+*(uint64_t*)row)
    num_rows++;

WTDMemTable::flush():
    append(_buffer, footer)
    write_to_disk(_buffer)
```

**Comment by decster** [comment]
>     // may missing key cols(partition cols or distributed cols), should fillin default value.
>     // If we want to support default values for each write txn, those default values need to 

If it's key column, fillin default value doesn't make sense, key column should be unique right?

About default value for non-key columns, it should be write to each Txn file, maybe in the footer PB?

**Comment by vagetablechicken** [comment]
> > ```
> > // may missing key cols(partition cols or distributed cols), should fillin default value.
> > // If we want to support default values for each write txn, those default values need to 
> > ```
> 
> If it's key column, fillin default value doesn't make sense, key column should be unique right?
> 
> About default value for non-key columns, it should be write to each Txn file, maybe in the footer PB?

My bad, I mix key columns with value columns. The key columns should be checked in calc_partition_and_tablet(), and if any key column is missing, we should return error(abort sending).

About default value for non-key columns, footer PB has `The schema part(ColumnMetaPB)`, so the default values for non-key columns are already in the file. I didn't point out it in Design, it'll be improved.

**Comment by morningman** [comment]
> key column should be unique right

I have a question that: What is "key column"? Is it a kind of primary key? Or just a partition or distribution column?

Does the "memory engine table" has primary key columns?

**Comment by decster** [comment]
> Does the "memory engine table" has primary key columns?

Yes, to support update/delete every row should have a primary key. This is equivalent to:

Agg table with agg key as primary key
Uniq table with uniq key as primary key
Even for the duplicate table,  a hidden seq id column is required to identify updates belonging to which row.

So a update like:

```
update t set col1 = col1+1 where xxx;
```

first scan rows with seqid and col1 and filter xxx, and then construct a writetxn with primary key and updated col1.

**Comment by decster** [comment]
Oh, initially memory engine only supports uniq table(which should be enough for typical use cases), and it can support (and probably unify) other table types later.

**Comment by morningman** [comment]
> Oh, initially memory engine only supports uniq table(which should be enough for typical use cases), and it can support (and probably unify) other table types later.

OK, I see~
